cvmfs/network/download.cc

```diff
@@ -1587,6 +1587,57 @@ bool DownloadManager::VerifyAndFinalize(const int curl_error, JobInfo *info) {
 }
 
 DownloadManager::~DownloadManager() {
+  // cleaned up fini
+  if (sharding_policy_.UseCount() > 0) {
+    sharding_policy_.Reset();
+  }
+  if (health_check_.UseCount() > 0) {
+    if (health_check_.Unique()) {
+      LogCvmfs(kLogDownload, kLogDebug, "Stopping healthcheck thread");
+      health_check_->StopHealthcheck();
+    }
+    health_check_.Reset();
+  }
+
+  if (atomic_xadd32(&multi_threaded_, 0) == 1) {
+    // Shutdown I/O thread
+    pipe_terminate_->Write(kPipeTerminateSignal);
+    pthread_join(thread_download_, NULL);
+    // All handles are removed from the multi stack
+    pipe_terminate_.Destroy();
+    pipe_jobs_.Destroy();
+  }
+
+  for (set<CURL *>::iterator i = pool_handles_idle_->begin(),
+       iEnd = pool_handles_idle_->end(); i != iEnd; ++i)
+  {
+    curl_easy_cleanup(*i);
+  }
+
+  delete pool_handles_idle_;
+  delete pool_handles_inuse_;
+  curl_multi_cleanup(curl_multi_);
+
+  delete header_lists_;
+  if (user_agent_)
+    free(user_agent_);
+
+  delete counters_;
+
+  if (opt_host_chain_) {
```

Comment:
No `NULL` check around `delete` (also further down). It's already done by `delete`.